packages/next/server/node-polyfill-fetch.js

```diff
@@ -1,23 +1,56 @@
-import fetch, {
-  Headers,
-  Request,
-  Response,
-} from 'next/dist/compiled/node-fetch'
-
 // Polyfill fetch() in the Node.js environment
+
 if (!global.fetch) {
-  const agent = ({ protocol }) =>
-    protocol === 'http:' ? global.__NEXT_HTTP_AGENT : global.__NEXT_HTTPS_AGENT
-  const fetchWithAgent = (url, opts, ...rest) => {
-    if (!opts) {
-      opts = { agent }
-    } else if (!opts.agent) {
-      opts.agent = agent
+  function getFetchImpl() {
+    return global.__NEXT_USE_UNDICI
+      ? require('next/dist/compiled/undici')
+      : require('next/dist/compiled/node-fetch')
+  }
+  // Due to limitation of global configuartion, we have to do this resolution at runtime
+  global.fetch = (...args) => {
+    const fetchImpl = getFetchImpl()
+
+    if (global.__NEXT_USE_UNDICI) {
+      // Undici does not support the `keepAlive` option,
+      // instead we have to pass a custom dispatcher
+      if (
+        !global.__NEXT_HTTP_AGENT_OPTIONS.keepAlive &&
+        !global.__NEXT_UNDICI_AGENT_SET
+      ) {
+        global.__NEXT_UNDICI_AGENT_SET = true
+        fetchImpl.setGlobalDispatcher(new fetchImpl.Agent({ pipelining: 0 }))
+      }
+      return fetchImpl.fetch(...args)
     }
-    return fetch(url, opts, ...rest)
+    const agent = ({ protocol }) =>
```

Comment:
This could be a problem because the agent is created for every `fetch()`. It should be reused, right?

Reply:
No, the Agent is created and assigned to `global.__NEXT_HTTP_AGENT` or `global.__NEXT_HTTPS_AGENT`. This function just returns the correct agent every time fetch is called.

But since we can't do that for the Undici agent we have to have all that `!global.__NEXT_UNDICI_AGENT_SET` logic.